src/data/setoid.lean

```diff
@@ -0,0 +1,603 @@
+/-
+Copyright (c) 2019 Amelia Livingston. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Amelia Livingston, Bryan Gin-ge Chen
+-/
+
+import data.quot data.set.lattice data.fintype order.order_iso order.galois_connection
+
+/-!
+# Equivalence relations
+
+The first section of the file defines the complete lattice of equivalence relations 
+on a type, results about the inductively defined equivalence closure of a binary relation, 
+and the analogues of some isomorphism theorems for quotients of arbitrary types.
+
+The second section comprises properties of equivalence relations viewed as partitions, and 
+a function "bell" that computes Bell numbers - the numbers of possible partitions of a finite 
+set - along with some preliminary results about "bell".
+
+## Implementation notes
+
+The function "rel" and lemmas ending in ' make it easier to talk about different 
+equivalence relations on the same type.
+
+The complete lattice instance for setoids could be defined by lifting the Galois
```

Comment:
I guess `lattice.copy` is designed for this use case. You could give your definitions of `le` and `inf`, and then prove all the complete-lattice-axioms using the Galois insertion. I don't know if it would save space.

Reply:
Thank you for all the help! I've applied all the comments in the latest commit, apart from this one. I've experimented with this approach using order.filter.basic as guidance and I don't think it would save space or clarity. I could remove that docstring, though?

Reply:
Maybe condense this discussion into your comment on line 25. And leave the code as it is.